### PR TITLE
[SPARK-7373][MESOS] Add docker support for launching drivers in mesos cluster mode.

### PR DESCRIPTION
**Pull request by tnachen**
Using the existing docker support for mesos, also enabling the mesos cluster mode scheduler to launch Spark drivers in docker images as well.

This also allows the executors launched by the drivers to be also in the same Docker image by passing  the docker settings.

### CODE REVIEW

**Comment by tnachen** [comment]
@andrewor14 PTAL, this is the last manually merged patch that we use to run DCOS Spark internally.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 1]
```diff
@@ -371,15 +371,30 @@ private[spark] class MesosClusterScheduler(
...
     val executorUri = desc.schedulerProperties.get("spark.executor.uri")
       .orElse(desc.command.environment.get("SPARK_EXECUTOR_URI"))
     val appArguments = desc.command.arguments.mkString(" ")
-    val cmd = if (executorUri.isDefined) {
+    val (executable, jar) = if (dockerDefined) {
+      val executorOpts = desc.schedulerProperties
+        .filterKeys(_.contains("docker"))
+        .map { case (k, v) =>
+        s"-D$k=$v"
+      }.mkString(" ")
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: style:
```
  .filterKeys(...)
  .map { case (k, v) => ... }
  .mkString(" ")
```

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 2]
```diff
@@ -371,15 +371,30 @@ private[spark] class MesosClusterScheduler(
...
+      // Pass docker options to executors
+      envBuilder.addVariables(
+        Variable.newBuilder()
+          .setName("SPARK_EXECUTOR_OPTS")
+          .setValue(executorOpts))
+      // Application jar is automatically downloaded in the mounted sandbox by Mesos,
+      // and the path to the mounted volume is stored in $MESOS_SANDBOX env variable.
+      ("./bin/spark-submit", s"$$MESOS_SANDBOX/${desc.jarUrl.split("/").last}")
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: sorry, maybe I'm missing something but why are there two dollar signs?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 3]
```diff
@@ -389,9 +404,9 @@ private[spark] class MesosClusterScheduler(
         }
       val cmdExecutable = new File(executorSparkHome, "./bin/spark-submit").getCanonicalPath
       val cmdJar = desc.jarUrl.split("/").last
-      s"$cmdExecutable ${cmdOptions.mkString(" ")} $cmdJar $appArguments"
+      (cmdExecutable, cmdJar)
     }
-    builder.setValue(cmd)
+    builder.setValue(s"$executable ${cmdOptions.mkString(" ")} $jar $appArguments")
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: in L373 up there can you just set `val cmdOptions = generateCmdOptions(desc).mkString(" ")`, so this is easier to read? I believe it's only used in one place so it's safe to do that.

**Comment by andrewor14** [comment]
@tnachen This looks fine. Once you address the minor comments I'll go ahead and merge this. I assume you've been testing this internally in production already.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala (hunk 2 above)
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: I want to have a literal $MESOS_SANDBOX, with string template $ will be evaluated so I need to escape with another $

**Comment by tnachen** [comment]
@andrewor14 yes this is what is running in our cluster right now, PTAL

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 4]
```diff
@@ -370,16 +370,30 @@ private[spark] class MesosClusterScheduler(
...
       Variable.newBuilder().setName("SPARK_EXECUTOR_OPTS").setValue(executorOpts))
-    val cmdOptions = generateCmdOption(desc)
+    val cmdOptions = generateCmdOption(desc).mkString(" ")
+    val dockerDefined = desc.schedulerProperties.contains("spark.mesos.executor.docker.image")
     val executorUri = desc.schedulerProperties.get("spark.executor.uri")
       .orElse(desc.command.environment.get("SPARK_EXECUTOR_URI"))
     val appArguments = desc.command.arguments.mkString(" ")
-    val cmd = if (executorUri.isDefined) {
+    val (executable, jar) = if (dockerDefined) {
+      val executorOpts = desc.schedulerProperties
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: `SPARK_EXECUTOR_OPTS` is already set to all properties in `desc.schedulerProperties` above. Do we need to set it for docker specific properties again? Doesn't it overwrite the setting?

Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: Ah yes you're right, I believe I added the pass all options later than this and it's a merge conflict problem. Thanks for catching this

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
retest this please. I believe the test failures are not related.

**Comment by andrewor14** [comment]
LGTM master 1.4 TYVM @tnachen.

**Pull request closed without merging**